returnn/datasets/postprocessing.py

```diff
@@ -0,0 +1,253 @@
+            When both functions are specified, the segment-level function is applied first to every segment, and the
+            results are passed to the iterator-level function.
+        :param map_outputs: Type and axis specification of the outputs of the mapping functions,
+            like extern_data and model_outputs.
+            To simplify the common case when no shapes change, this value can be left unspecified. The dataset then
+            assumes the same data layout as returned by the wrapped dataset.
+        :param _meta_info_cache: internal usage
+        :param kwargs: see :class:`CachedDataset2`, :class:`Dataset`
+        """
+        super().__init__(**kwargs)
+
+        assert (
+            self.seq_ordering == "default"
+        ), f"specify seq_ordering in wrapped dataset, not in {self.__class__.__name__}"
+        assert (
+            map_seq is not None or map_seq_stream is not None
+        ), "need to either define map_seq or map_seq_stream functions"
+
+        self._dataset_def = dataset
+        self._map_seq = map_seq
+        self._map_seq_stream = map_seq_stream
+        self._map_outputs = map_outputs
+
+        self._dataset: Optional[Dataset] = None
+        self._data_keys: Optional[List[str]] = None
+        self._data_iter: Optional[Iterator[Tuple[int, TensorDict]]] = None
+        self._default_input: Optional[str] = None
+        self._dim_cache: Dict[str, Tuple[List[Dim], Optional[Dim]]] = {}
+
+        if _meta_info_cache:
+            self._default_input = _meta_info_cache["_default_input"]
+            self._dim_cache = _meta_info_cache["_dim_cache"]
+            self._estimated_num_seqs = _meta_info_cache["_estimated_num_seqs"]
+            self.labels = _meta_info_cache["labels"]
+            self.num_inputs = _meta_info_cache["num_inputs"]
+            self.num_outputs = _meta_info_cache["num_outputs"]
+
+    def initialize(self):
+        """init"""
+        self._lazy_init_dataset()
+        super().initialize()
+
+    @property
+    def _meta_info_cache(self) -> Optional[Dict[str, Any]]:
+        if self.num_outputs is None:
+            return None
+        return {
+            "_default_input": self._default_input,
+            "_dim_cache": self._dim_cache,
+            "_estimated_num_seqs": self._estimated_num_seqs,
+            "labels": self.labels,
+            "num_inputs": self.num_inputs,
+            "num_outputs": self.num_outputs,
+        }
+
+    def _lazy_init_dataset(self):
+        if self._dataset is not None:
+            return
+
+        self._dataset = init_dataset(self._dataset_def, parent_dataset=self)
+        self.labels = self._dataset.labels
+        self._default_input = (
+            "data" if not self._map_outputs or "data" in self._map_outputs else next(self._map_outputs.keys())
+        )
+        self._estimated_num_seqs = self._dataset.estimated_num_seqs
+
+        if self._map_outputs is not None:
+            tdict = TensorDict(self._map_outputs)
```

Comment:
Better also store this in the attrib.

Reply:
The `Dim`s are already created here.

Reply:
But only if `_map_outputs` is defined in `extern_data`-shape/type. If not, and we want to use the shape from the sub dataset we need to transform the sub-dataset's `num_output` into a form that is usable to initialize the `TensorDict` with. Not sure this is a good way to go, unless we switch over _all_ datasets to `TensorDict` now?

Reply:
No, `Dim`s are always created.

Reply:
> always 

I don't understand why, the code initializes the `TensorDict` there only when `_map_outputs` is defined. Where is the other place?

Reply:
I would actually suggest that if `_map_outputs` is not defined, you also create a `TensorDict` here, but derived from the subdataset. Then you have the code more consistent and simple and all `Dim`s already created here.

Reply:
Yes, okay, then I agree this would be nice. 

> If not, and we want to use the shape from the sub dataset we need to transform the sub-dataset's num_output into a form that is usable to initialize the TensorDict with.

I don't like that we need to transform the info from `self._dataset.num_outputs` though. Is there even enough info in there?

Reply:
> I don't understand why, the code initializes the `TensorDict` there only when `_map_outputs` is defined. Where is the other place?

I meant for the case that `map_outputs` is defined. Then you even *must* reuse the `Dim`s from there.

But I was also suggesting to create the `TensorDict` just in all cases right here. Then the code is also simpler.

Reply:
> I don't like that we need to transform the info from `self._dataset.num_outputs` though. Is there even enough info in there?

But that's exactly what you do already right now in `_make_tensor`, using `get_data_dim` and `get_data_shape`? I just say to move that logic right here, so you always have the template `TensorDict` already ready, and you can just get the `Dim`s from there.

Reply:
Yes, it's probably all wrong already. Imagine the case when one PostprocessingDataset wraps the other, the outer one then accesses `num_outputs` of the inner one, and thereby loses the `Dim` information potentially declared in the inner `map_outputs`-dict.

Reply:
But `num_outputs` does not expose any `Dim`s?